File: 101-symmetric-tree/symmetric-tree.java
class Solution {
    public boolean isSymmetric(TreeNode root) {

        return root == null || isMirror(root.left, root.right);
    }

    private boolean isMirror(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        }
        
        // If only one is null, they are not symmetric
        if (left == null || right == null) {
            return false;
        }

        // Check if values are equal and subtrees are mirrors
        return (left.val == right.val) &&
               isMirror(left.left, right.right) &&
               isMirror(left.right, right.left);
    }
}

File: 101-symmetric-tree/symmetric-tree.cpp
class Solution {
public:
    bool solve(TreeNode* r1, TreeNode* r2) {
        if (!r1 && !r2) return true;
        if (!r1 || !r2) return false;

        if (r1->val != r2->val) return false;

        return solve(r1->left, r2->right) && solve(r1->right, r2->left);
    }

    bool isSymmetric(TreeNode* root) {
        if (!root) return true;

        return solve(root->left, root->right);
    }
};
